notes-0319-inductive-types.tex

```diff
@@ -0,0 +1,228 @@
+For example, consider the vector type of lists of a specified length $Vec_n(A)$ for $n : \N$ generated by the constructors $[] : Vec_0(A)$ and $cons : \prod_{n : \N} A \rightarrow Vec_n(A) \rightarrow Vec_{suc\ n}(A)$.
+Note that the cons constructor mixes the value of $n$, so we say that $n$ is an index of the type family $Vec_{(\cdot)}(\cdot)$.
+On the other hand, $A$ is still a type parameter for this type family, for the same reason it was a parameter of the list type.
+
+
+\section{Higher Inductive Types}
+Besides the inductive types defined above, we would also like to define types like the circle:
+  $\Sp$ with constructors $base : \Sp$ and $loop : Id(base ; base)$.
+
+This definition of the circle is more complicated than other types, because we've introduced a path, not just new elements.
+Therefore this is a \textbf{higher inductive type}.
+Writing down the rules for this type is more work than for regular inductive types, but the introduction rules are still as we'd expect:
+
+\begin{multicols}{3}
+  \noindent
+  \begin{equation*}
+    \begin{prooftree}
+      \infer0[$\Sp$-Formation] {
+        \Sp : U
+      }
+    \end{prooftree}
+  \end{equation*}
+  \begin{equation*}
+    \begin{prooftree}
+      \infer0[base-Intro] {
+        base : \Sp
+      }
+    \end{prooftree}
+  \end{equation*}
+  \begin{equation*}
+    \begin{prooftree}
+      \infer0[loop-Intro] {
+        loop : Id_{\Sp}(base; base)
+      }
+    \end{prooftree}
+  \end{equation*}
+\end{multicols}
+
+In order to make sense of the uniqueness rules, we'll first have to define a dependent version of $ap$.
+While $ap$ allows us to apply a function to a path to get a path in the co-domain, what we need now is a way to apply a dependent function to a path to get a path in the total space, where we think of a path in the total space as a path over the path in the base space.
+
+After doing that we are able to give the elimination rule for this type:
+
+\begin{equation*}
+  \begin{prooftree}
+    \hypo{x : \Sp \rightarrow C : U}
+    \hypo{M_{base} : C[base/x]}
+    \hypo{M_{loop} : base =_{loop} base}
+    \hypo{N : \Sp}
+    \infer4[S1-Elimination] {
+      elim_{\Sp}[x.C](M_{base}; M_{loop}; N) : C[N/x]
+    }
+  \end{prooftree}
+\end{equation*}
+
+As well as a pair of computation rules for the two constructors:
+
+\begin{equation*}
+  \begin{prooftree}
+    \hypo{x : \Sp \rightarrow C : U}
+    \hypo{M_{base} : C[base/x]}
+    \hypo{M_{loop} : base =_{loop} base}
+    \infer3[S1-Elimination] {
+      elim_{\Sp}[x.C](M_{base}; M_{loop}; base) \equiv M_{base}
+    }
+  \end{prooftree}
+\end{equation*}
+\begin{equation*}
+  \begin{prooftree}
+    \hypo{x : \Sp \rightarrow C : U}
+    \hypo{M_{base} : C[base/x]}
+    \hypo{M_{loop} : base =_{loop} base}
+    \infer3[S1-Elimination] {
+      apd(\lambda x . elim_{\Sp}[x.C](M_{base}; M_{loop}; N); loop) \equiv M_{loop} 
+    }
+  \end{prooftree}
+\end{equation*}
+
```

Comment:
This is already good, but perhaps you could mention the S1-algebra to echo the previous part about algebras.

Reply:
I think a definition of path-over-path and apd would make the section more complete.